### PR TITLE
Value mapping

### PR DESCRIPTION
**Pull request by veerajsk**

### CODE REVIEW

**Comment by veerajsk** [comment]
@bjoernschroeder is working on the unit tests, this PR is for the engine changes only

**Review by Dale-Koenig — APPROVED** [review]

**Comment by Dale-Koenig** [comment]
https://github.com/rapyuta-robotics/alica/pull/458
At least input mappings are working

File: alica_tests/src/behaviours/TestParameterPassingBehaviour.cpp [hunk 1]
```diff
@@ -13,20 +13,26 @@ TestParameterPassingBehaviour::TestParameterPassingBehaviour(BehaviourContext& c
 TestParameterPassingBehaviour::~TestParameterPassingBehaviour() {}
 void TestParameterPassingBehaviour::run()
 {
-    setSuccess();
+    if (isSuccess()) {
+        return;
+    }
+
+    LockedBlackboardRO gb(*getGlobalBlackboard());
+    if (gb.get<int64_t>("param1") == 5 && gb.get<int64_t>("param2") == 7) {
```
Comment (veerajsk) [inline alica_tests/src/behaviours/TestParameterPassingBehaviour.cpp]: can do this in initialise parameters itself. There's no reason to set on the gb & then verify from there. Can directly verify from bb.get

File: alica_tests/src/behaviours/TestParameterPassingBehaviour.cpp [hunk 2]
```diff
@@ -13,20 +13,26 @@ TestParameterPassingBehaviour::TestParameterPassingBehaviour(BehaviourContext& c
...
+    LockedBlackboardRW bb(*getBlackboard());
     bb.set<int64_t>("behaviorKey", 2);
-    wm->passedParameters["behaviorKey"] = bb.get<int64_t>("behaviorKey");
 
     auto value = bb.get<int64_t>("behaviorInputKey");
     if (value == 5) {
-        wm->passedParameters["behaviorInputKey"] = value;
+        gb.set("param1", value);
     } else if (value == 7) {
-        wm->passedParameters["behaviorSecondInputKey"] = value;
```
Comment (veerajsk) [inline alica_tests/src/behaviours/TestParameterPassingBehaviour.cpp]: I feel that refactoring of existing tests should be done in a separate PR because otherwise, it makes the PR less focused, big & hard to review. Can you please only include the changes required for value mapping here?

File: alica_tests/src/behaviours/ValueMappingTestBeh.cpp [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+void ValueMappingTestBeh::run()
+{
+    if (isSuccess()) {
+        return;
+    }
+
+    LockedBlackboardRO bb(*getBlackboard());
+    if (bb.get<int>("inputInt") == -9 && bb.get<uint64_t>("inputUint") == 5 && bb.get<bool>("inputBool") == true &&
```
Comment (veerajsk) [inline alica_tests/src/behaviours/ValueMappingTestBeh.cpp]: would be nice to check & store the reason for failure on the gb (including catching exceptions thrown by get)

File: alica_tests/src/plans/TestBlackboardPlan.cpp [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+{
+TestBlackboardPlan::TestBlackboardPlan(PlanContext& context)
+        : AlicaTestsPlan(context)
+{
+}
+
+bool ValueMappingConditionTestState2BlackboardTestSuccessState(const Blackboard* input, const RunningPlan* rp, const Blackboard* gb)
+{
+    alica::LockedBlackboardRO bb(*input);
+    return bb.get<bool>("inputBool") == true && bb.get<int64_t>("inputInt") == 17 && bb.get<std::string>("inputString") == "test" &&
```
Comment (veerajsk) [inline alica_tests/src/plans/TestBlackboardPlan.cpp]: same here, can store a value on the gb containing the reason for failure & print that out in the test

File: alica_tests/src/plans/ValueMappingPlansPlan.cpp [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+{
+ValueMappingPlansPlan::ValueMappingPlansPlan(PlanContext& context)
+        : AlicaTestsPlan(context)
+{
+}
+
+bool ValueMappingPlansState2ValueMappingPlansSuccessState(const Blackboard* input, const RunningPlan* rp, const Blackboard* gb)
+{
+    alica::LockedBlackboardRO bb(*input);
+    return bb.get<bool>("inputBool") == true && bb.get<int64_t>("inputInt") == 16 && bb.get<std::string>("inputString") == "test" &&
```
Comment (veerajsk) [inline alica_tests/src/plans/ValueMappingPlansPlan.cpp]: this is not really checking value mapping to a plan, rather it is checking value mapping to a plan followed by key mapping to a condition. So if a failure occurs its hard to tell if it is due to value mapping or key mapping. Instead I think you should be checking in the plan init if the keys got mapped correctly

File: alica_tests/src/test/etc/plans/AuthorityTest.pml [hunk 6]
```diff
@@ -77,7 +78,7 @@
       "comment": "",
       "confAbstractPlanWrappers": [
         {
-          "abstractPlan": "behaviours/EmptyBehaviour.beh#1625610857563",
+          "abstractPlan": "EmptyBehaviour.beh#1625610857563",
```
Comment (veerajsk) [inline alica_tests/src/test/etc/plans/AuthorityTest.pml]: Again, would prefer to do this in a separate PR

File: alica_tests/src/test/test_blackboard.cpp [hunk 7]
```diff
@@ -34,68 +36,161 @@ class NonConstructable
...
-            LockedBlackboardRW(ae->editGlobalBlackboard()).get<std::shared_ptr<alicaTests::TestWorldModel>>("worldmodel");
-    EXPECT_EQ(wm->passedParameters["behaviorInputKey"], 5);       // Value set in first behavior call
-    EXPECT_EQ(wm->passedParameters["behaviorSecondInputKey"], 7); // Value set in second behavior call
+    // Transition to the plan corresponding to this test case
+    ASSERT_TRUE(_tc->setTransitionCond("TestMasterPlan", "ChooseTestState", "BlackboardTestState")) << _tc->getLastFailure();
+    STEP_UNTIL(_tc, _tc->getActivePlan("TestBlackboardPlan"));
+    auto plan = _tc->getActivePlan("TestBlackboardPlan");
+    ASSERT_NE(plan, nullptr) << _tc->getLastFailure();
+
+    ASSERT_TRUE(_tc->setTransitionCond("TestBlackboardPlan", "SelectBlackboardTestState", "JsonTwoBehaviorKeyMappingTestState")) << _tc->getLastFailure();
```
Comment (veerajsk) [inline alica_tests/src/test/test_blackboard.cpp]: I don't understand what's going on with these testjson* tests or the empty states in TestBlackboard Plan

**Review by veerajsk — COMMENTED** [review]
@bjoernschroeder I think this PR should be broken up into multiple PRs & I am confused with some of the test plans. It would be easy to review if
- put alica_test changes in a separate PR & revert the last commit. This is needed because I can't request changes etc. to this PR because github says I'm the author
- the alica_tests etc folder is generated with the updated designer in a different PR
- Existing tests & plan changes are done in another PR
- we can discuss what's going on with the empty states in the TestBlackboardPlan

File: alica_tests/src/test/etc/plans/ConfigurationTestPlan.pml [hunk 8]
```diff
@@ -29,6 +29,7 @@
   ],
   "frequency": 0,
   "id": 1588060981661,
+  "implementationName": "",
```
Comment (veerajsk) [inline alica_tests/src/test/etc/plans/ConfigurationTestPlan.pml]: engine PR is not yet merged, so this shouldn't be here imo

**Review by bjoernschroeder — APPROVED** [review]
